accepted/0000-overrides.md

```diff
@@ -0,0 +1,492 @@
+    - apply `O'` to `x`, due to matching key.  Duplicate version with same
+      overrides object `O'` in root already, dedupe against existing dep.
+    - apply `O` to `i`.  Duplicate version with same overrides object `O` in
+      root already, dedupe against existing dep.
+- Resolve dependencies of `x>a` (in `node_modules/x/node_modules`), `(b,c)`
+    - apply `O''` to `b`.  Existing version in root `node_modules` has a
+      different overrides object, so place at
+      `node_modules/x/node_modules/b`.
+    - `c` resolves to `1.3.0`, but override applies.  Flag `x>a` and `x` as
+      having relevant overrides.  Different version of `c` in top level, so
+      place at `node_modules/x/node_modules/c`.
+- Resolve dependencies of `x>a>b` (in `node_modules/x/node_modules`), `(i, c)`
+    - override `c` from `1.3.0` to `1.2.3` due to override match.  Same
+      version present already in `node_modules/x/node_modules/c`, so dedupe
+      against existing dep.  Flag `x>a>b` overrides as relevant.
+    - Apply `O''` to `i`.  Existing version at
+      `node_modules/x/node_modules/i` has overrides object `O'`, so place
+      at `node_modules/x/node_modules/b/node_modules/i`.
+- Resolve dependencies of `x>a>c`, `()`
+    - empty set, nothing to do
+- Resolve dependencies of `x>a>b>i` (j)
+    - Apply `O''` to `j`.  Existing version at
+      `node_modules/x/node_modules/j` has overrides object `O'`, so place
+      at `node_modules/x/node_modules/b/node_modules/j`.
+- Resolve dependencies of `x>a>b>i>j` ()
+    - empty set, nothing to do
+
+If we stopped and reified at this point, we'd have the following physical
+tree:
+
+```
+root
++-- a
++-- b
++-- c (1.3.0)
++-- i (*)
++-- j (*)
++-- y
++-- x
+    +-- a
+    +-- b
+    |   +-- i (*)
+    |   +-- j (*)
+    +-- c (1.2.3)
+    +-- i (*)
+    +-- j (*)
+```
+
+At no point was the `overridesRelevant` flag set for the nodes marked
+`(*)`, so they are free to be deduped.
+
+- For each dep in the tree, where `dep.overridesRelevant` is false, and
+  `overrides` is an object
+    - set `dep.overrides = null`
+    - add it to a list
+- Sort list in ascending order by `dep.depth`, then alphabetically by
+  `dep.name`.
+- For each item in the list, search the tree from the dep's parent for any
+  packages by the same name which can be safely deduplicated against them
+  (meaning: a compatible match for all edges resolving to the node, no
+  incompatible shadowed dependencies, having no relevant `overrides` object
+  attached), and remove those from the tree.
+
+The resulting physical tree after deduplication would be:
+
+```
+root
++-- a
++-- b
++-- c (1.3.0)
++-- i (*)
++-- j (*)
++-- y
++-- x
+    +-- a
+    +-- b
+    +-- c (1.2.3)
+```
+
+Only nodes `a`, `b`, and `c` under `x` are duplicated, because they are
+subject to a relevant `overrides` setting.
+
+## Questions and Bikeshedding
+
+- Should `bundleDependencies` and `shrinkwrap` dependencies be subject to
```

Comment:
I would not expect bundleDeps to be affected, since the whole point of them is "npm is not allowed to touch bundled deps". Not sure what you mean about "shrinkwrap deps"?

Reply:
Dependencies of a module that includes a `npm-shrinkwrap.json` file.  Again, the whole point of them is "npm is not allowed to use different deps than the tree specified in the shrinkwrap", so in most ways, they're treated just like bundled deps (albeit, bundled deps that we still have to fetch and unpack).

Reply:
It seems to me like that shouldn't be touched either.